table/block_based/block.cc

```diff
@@ -607,12 +638,13 @@ bool IndexBlockIter::ParseNextIndexKey() {
 // shared_size is 0, which included the first entry in each restart point.
 // Otherwise the format is delta-size = block handle size - size of last block
 // handle.
-void IndexBlockIter::DecodeCurrentValue(uint32_t shared) {
-  Slice v(value_.data(), data_ + restarts_ - value_.data());
+void IndexBlockIter::DecodeCurrentValue(bool is_shared) {
+  auto limit = block_->offset(block_->block_limit());
+  Slice v(value_.data(), limit - value_.data());
   // Delta encoding is used if `shared` != 0.
   Status decode_s __attribute__((__unused__)) = decoded_value_.DecodeFrom(
       &v, have_first_key_,
-      (value_delta_encoded_ && shared) ? &decoded_value_.handle : nullptr);
+      (value_delta_encoded_ && is_shared) ? &decoded_value_.handle : nullptr);
```

Comment:
We have seen a couple of times in stress test violation of the final assertion in DecodeEntry. (Internal ref T100361844.) In looking into that, I noticed that our test suite never reaches this point in the code with  `global_seqno_state_ != nullptr && !value_delta_encoded_`. Since you are refactoring this code, it would be nice to add coverage if we can, potentially find any bugs old and/or new.